server/src/main/java/io/spine/server/commandbus/TenantAwareSystemGateway.java

```diff
@@ -0,0 +1,140 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.server.commandbus;
+
+import com.google.protobuf.Message;
+import io.spine.core.TenantId;
+import io.spine.server.tenant.TenantAwareOperation;
+import io.spine.system.server.SystemGateway;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * A {@link SystemGateway} which works for a specified tenant.
+ *
+ * <p>A system command is {@linkplain #postCommand(Message) posted} within
+ * a {@link TenantAwareOperation} with the given tenant set.
+ *
+ * <p>Any call to this gateway is delegated to another instance of {@link SystemGateway}
+ * {@linkplain Builder#atopOf(SystemGateway) passed} on construction.
+ *
+ * @author Dmytro Dashenkov
+ * @see TenantAwareOperation
+ */
+final class TenantAwareSystemGateway implements SystemGateway {
+
+    private final SystemGateway delegate;
+    private final TenantId tenantId;
+
+    private TenantAwareSystemGateway(Builder builder) {
+        this.delegate = builder.delegate;
+        this.tenantId = builder.tenantId;
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * <p>On an instance of {@code TenantAwareSystemGateway}, posts the given system command for
+     * the specified tenant.
+     */
+    @Override
+    public void postCommand(Message systemCommand) {
+        Runnable action = () -> delegate.postCommand(systemCommand);
+        TenantAwareOperation operation = new Operation(tenantId, action);
+        operation.execute();
+    }
+
+    /**
+     * A {@link TenantAwareOperation} which executes the given action for a specific tenant.
+     */
+    private static final class Operation extends TenantAwareOperation {
+
+        private final Runnable action;
+
+        private Operation(TenantId tenantId, Runnable action) {
+            super(tenantId);
+            this.action = action;
+        }
+
+        @Override
+        public void run() {
+            action.run();
+        }
+    }
+
+    /**
+     * Creates a new instance of {@code Builder} for {@code TenantAwareSystemGateway} instances.
+     *
+     * @return new instance of {@code Builder}
+     */
+    public static Builder create() {
+        return new Builder();
+    }
+
+    /**
+     * A builder for the {@code TenantAwareSystemGateway} instances.
+     */
+    static final class Builder {
+
+        private SystemGateway delegate;
+        private TenantId tenantId;
+
+        /**
+         * Prevents direct instantiation.
+         */
+        private Builder() {
+        }
+
+        /**
+         * Specifies the {@link SystemGateway} to delegate calls to.
+         *
+         * @param gateway the delegate {@link SystemGateway}
+         */
+        Builder atopOf(SystemGateway gateway) {
```

Comment:
How do you know that the builder works? Please have a test for the builder.

Reply:
Done.